src/3d/qgscameracontroller.cpp

```diff
@@ -430,40 +556,152 @@ void QgsCameraController::zoom( float factor )
   updateCameraFromPose();
 }
 
+void QgsCameraController::handleTerrainNavigationWheelZoom()
+{
+  if ( !mDepthBufferIsReady )
+    return;
+
+  if ( !mZoomPointCalculated )
+  {
+    double depth = Qgs3DUtils::decodeDepth( mDepthBufferImage.pixelColor( mMousePos.x(), mMousePos.y() ) );
+    mZoomPoint = Qgs3DUtils::screenPointToWorldPos( mMousePos, depth, mViewport.size(), mCameraBeforeZoom );
+    mZoomPointCalculated = true;
+  }
+
+  float f = mCumulatedWheelY / ( 120.0 * 24.0 );
+
+  double dist = ( mZoomPoint - mCameraBeforeZoom->position() ).length();
+  dist -= dist * f;
+
+  // First transformation : Shift camera position and view center and rotate the camera
+  {
+    QVector3D shiftVector = mZoomPoint - mCamera->viewCenter();
+
+    QVector3D newViewCenterWorld = camera()->viewCenter() + shiftVector;
+
+    mCameraPose.setDistanceFromCenterPoint( dist );
+    mCameraPose.setCenterPoint( newViewCenterWorld );
+    updateCameraFromPose();
+  }
+
+  // Second transformation : Shift camera position back
+  {
+    QgsRay3D ray = Qgs3DUtils::rayFromScreenPoint( QPoint( mMousePos.x(), mMousePos.y() ), mViewport.size(), mCamera );
+    QVector3D clickedPositionWorld = ray.origin() + dist * ray.direction();
+
+    QVector3D shiftVector = clickedPositionWorld - mCamera->viewCenter();
+
+    QVector3D newViewCenterWorld = camera()->viewCenter() - shiftVector;
+    QVector3D newCameraPosition = camera()->position() - shiftVector;
+
+    mCameraPose.setDistanceFromCenterPoint( ( newViewCenterWorld - newCameraPosition ).length() );
+    mCameraPose.setCenterPoint( newViewCenterWorld );
+    updateCameraFromPose();
+  }
+}
+
 void QgsCameraController::onWheel( Qt3DInput::QWheelEvent *wheel )
 {
   switch ( mCameraNavigationMode )
   {
     case QgsCameraController::WalkNavigation:
     {
-      const float scaling = ( ( wheel->modifiers() & Qt::ControlModifier ) ? 0.1f : 1.0f ) / 1000.f;
+      const float scaling = ( ( wheel->modifiers() & Qt::ControlModifier ) != 0 ? 0.1f : 1.0f ) / 1000.f;
       setCameraMovementSpeed( mCameraMovementSpeed + mCameraMovementSpeed * scaling * wheel->angleDelta().y() );
       break;
     }
 
     case TerrainBasedNavigation:
     {
-      const float scaling = ( ( wheel->modifiers() & Qt::ControlModifier ) ? 0.1f : 1.0f ) / 1000.f;
-      float dist = mCameraPose.distanceFromCenterPoint();
-      dist -= dist * scaling * wheel->angleDelta().y();
-      mCameraPose.setDistanceFromCenterPoint( dist );
-      updateCameraFromPose();
+
+      const float scaling = ( ( wheel->modifiers() & Qt::ControlModifier ) != 0 ? 0.5f : 5.f );
+
+      // Apparently angleDelta needs to be accumulated
+      // see: https://doc.qt.io/qt-5/qwheelevent.html#angleDelta
+      mCumulatedWheelY += scaling * wheel->angleDelta().y();
+
+      if ( !mIsInZoomInState )
+      {
+        if ( !mCameraBeforeZoom )
+          mCameraBeforeZoom = new Qt3DRender::QCamera;
```

Comment:
This object and the other QCamera objects do not seem to get deleted anywhere - maybe best to set their parent to QgsCameraController when they are created.

Also, I would suggest to simply create mCameraBeforeZoom and other QCameras in camera controller constructor - the extra cost is minimal and it will make me less nervous worrying that mCameraBeforeZoom may be a null pointer at some point :-)

Reply:
I implemented your suggession. Thanks for the reviews!